### ISSUE TITLE
hasNextPage / hasPreviousPage not respecting the Relay spec fully

### ISSUE THREAD

**Issue by jasonbahl** [wp-graphql/wp-graphql#594]
The GraphQL Relay spec states the following for `hasPreviousPage`:

HasPreviousPage(allEdges, before, after, first, last)
- If last is set:
   - Let edges be the result of calling ApplyCursorsToEdges(allEdges, before, after).
   - If edges contains more than last elements return true, otherwise false.
- If after is set:
   - If the server can efficiently determine that elements exist prior to after, return true.
- Return false.

## Currently, WPGraphQL is skipping:

```
- If after is set:
   - If the server can efficiently determine that elements exist prior to after, return true.
```

I think we should be able to efficiently determine if there are previous items so that we can handle this case appropriately. 

The same is the case with `hasNextPage`:

HasNextPage(allEdges, before, after, first, last)
- If first is set:
   - Let edges be the result of calling ApplyCursorsToEdges(allEdges, before, after).
   - If edges contains more than first elements return true, otherwise false.
- If before is set:
   - If the server can efficiently determine that elements exist following before, return true.
- Return false.

Currently, we're not respecting:

```
- If before is set:
   - If the server can efficiently determine that elements exist following before, return true.
```

-----

Thanks for pointing this out @MoOx!

**Comment by MoOx** [comment]
Tell me if I can help (even if I didn't do any php for ages).
I could also be a beta tester if necessary :)

**Comment by jasonbahl** [comment]
@MoOx even if you don't want to write tests, you _could_ write some client code, in something like CodeSandbox.io, write it up as you'd expect it to work, and we can use that as the "test", then someone with more PHP experience can convert that into actual Unit/Integration tests

**Comment by agyimr** [comment]
I am also struggling to get over this. I wanted to use hasNextPage and hasPreviousPage to allow the user to press the buttons "Next" and "Previous". But currently if I set "last" or "first" the other one (in the opposite direction I am paginating) resolves to false and the user can not go back. 

Any suggested workarounds until this gets officially supported? :) 

Keep up the good work!

**Comment by let00** [comment]
I can take a look at this. My current thoughts are to get the offset from the after cursor, and then check if the offset corresponds to a valid object ID. Assuming that the offset of the cursor is the ID of the node at the tail of the edge. Would this be a good approach?

**Comment by jasonbahl** [comment]
@let00 ya, I think that should work.

**Comment by jasonbahl** [comment]
closed by #1111
